iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp

```diff
@@ -0,0 +1,228 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_utils/cxx/function_ref.hpp"
+#include "test.hpp"
+
+using namespace ::testing;
+using namespace iox::cxx;
+
+int freeFunction()
+{
+    return 42 + 42;
+}
+
+class Functor
+{
+  public:
+    int operator()()
+    {
+        return 73;
+    }
+};
+
+struct ComplexType
+{
+    char a;
+    int b;
+    float c;
+
+    bool operator==(const ComplexType& rhs) const
+    {
+        if (a == rhs.a && b == rhs.b && c == rhs.c)
+        {
+            return true;
+        }
+        else
+        {
+            return false;
+        }
+    }
+};
+
+ComplexType returnComplexType(ComplexType foo)
+{
+    return foo;
+}
+
+class function_refTest : public Test
+{
+  public:
+    void SetUp() override
+    {
+    }
+
+    void TearDown() override
+    {
+    }
+
+    int foobar()
+    {
+        return 4273;
+    }
+
+    uint8_t m_iter{0};
+};
+
+using function_refDeathTest = function_refTest;
+
+TEST_F(function_refTest, CreateEmpty)
+{
+    function_ref<void()> sut;
+    EXPECT_FALSE(sut);
+}
+
+TEST_F(function_refTest, CreateWithNullptr)
+{
+    function_ref<void()> sut(nullptr);
+    EXPECT_FALSE(sut);
+}
+
+TEST_F(function_refDeathTest, CreateEmptyLeadsToTermination)
+{
+    function_ref<void()> sut;
+    EXPECT_DEATH(sut(), ".*");
+}
+
+TEST_F(function_refTest, CreateEmptyAndAssign)
+{
+    auto lambda = [] {};
+    function_ref<void()> sut;
+    sut = lambda;
+    EXPECT_TRUE(sut);
+}
+
+TEST_F(function_refTest, CreateAndCopy)
+{
+    auto lambda = []() -> int { return 42; };
+    function_ref<int()> sut1(lambda);
+    function_ref<int()> sut2{sut1};
+    EXPECT_TRUE(sut2);
+    EXPECT_THAT(sut2(), Eq(42));
+}
+
+TEST_F(function_refTest, CreateAndCopyAssign)
+{
+    auto lambda = []() -> int { return 42; };
+    function_ref<int()> sut2;
+    {
+        function_ref<int()> sut1(lambda);
+        EXPECT_THAT(sut1(), Eq(42));
+        EXPECT_FALSE(sut2);
+        sut2 = sut1;
+    }
+    EXPECT_TRUE(sut2);
+    EXPECT_THAT(sut2(), Eq(42));
+}
+
+TEST_F(function_refTest, CreateAndMove)
+{
+    auto lambda = []() -> int { return 42; };
+    function_ref<int()> sut1{lambda};
+    function_ref<int()> sut2(std::move(sut1));
+    EXPECT_TRUE(sut2);
+    EXPECT_THAT(sut2(), Eq(42));
+    EXPECT_FALSE(sut1);
+}
+
+TEST_F(function_refTest, CreateAndMoveAssign)
+{
+    auto lambda1 = []() -> int { return 42; };
+    auto lambda2 = []() -> int { return 73; };
+    function_ref<int()> sut1{lambda1};
+    {
+        function_ref<int()> sut2{lambda2};
+        sut1 = std::move(sut2);
+    }
+    EXPECT_TRUE(sut1);
+    EXPECT_THAT(sut1(), Eq(73));
+}
+
+TEST_F(function_refTest, CreateAndSwap)
```

Comment:
We should also test swapping valid-invalid function_refs

Reply:
Added